ambuild2/frontend/v2_2/cpp/detect.py

```diff
@@ -0,0 +1,397 @@
+            return None
+
+        necessary_tools = ['cl.exe', 'rc.exe', 'lib.exe']
+        tools, _ = FindToolsInEnv(env, necessary_tools)
+        for tool in necessary_tools:
+            if tool not in tools:
+                util.con_err(util.ConsoleRed, "Could not find {} for {}".format(tool, bat_file))
+                return None
+
+        cc, _ = self.run_compiler('CC', 'cl', 'msvc', env, abs_path = tools['cl.exe'])
+        if not cc:
+            return None
+        cxx, _ = self.run_compiler('CXX', 'cl', 'msvc', env, abs_path = tools['cl.exe'])
+        if not cxx:
+            return None
+
+        # We use tuples here so the data is hashable without going through Pickle.
+        tool_list = (
+            ('cl', tools['cl.exe']),
+            ('rc', tools['rc.exe']),
+            ('lib', tools['lib.exe']),
+        )
+        env_data = {
+            ('env_cmds', env_cmds),
+            ('tools', tool_list),
+        }
+        return self.create_cli(cxx.vendor, cc.argv, cxx.argv, env_data)
+
+    def find_default_compiler(self):
+        candidates = []
+        if self.host_.platform == 'windows':
+            candidates.append(kMsvcTuple)
+        candidates.extend([kClangTuple, kGccTuple, kIntelTuple])
+
+        for cc_cmd, cxx_cmd, cc_family in candidates:
+            cc = self.find_compiler('CC', cc_cmd, cc_family)
+            if cc is None:
+                continue
+            cxx = self.find_compiler('CXX', cxx_cmd, cc_family)
+            if cxx is not None:
+                return cc, cxx
+
+        raise CompilerNotFoundException('Unable to find a suitable C/C++ compiler')
+
+    def find_compiler(self, mode, cmd, cc_family = None):
+        families = []
+        if "EMSDK" in os.environ and cmd[:2] == 'em':
+            families.append('emscripten')
+        if cc_family is not None:
+            families.append(cc_family)
+        else:
+            # On Windows, check for MSVC compatibility before checking GCC.
+            if self.host_.platform == 'windows':
+                families.append('msvc')
+            families.append('gcc')
+
+        for family in families:
+            compiler, e = self.run_compiler(mode, cmd, family)
+            if compiler is not None:
+                return compiler
+            if isinstance(e, CompilerNotFoundException):
+                # No reason to keep trying the same command.
+                break
+
+        return None
+
+    def run_compiler(self, mode, cmd, assumed_family, env = None, abs_path = None):
+        flags = shlex.split(os.environ.get('CFLAGS', ''))
+        if self.cflags_:
+            flags.extend(self.cflags_)
+        if mode == 'CXX':
+            flags.extend(shlex.split(os.environ.get('CXXFLAGS', '')))
+            if self.cxxflags_:
+                flags.extend(self.cxxflags_)
+
+        try:
+            return VerifyCompiler(flags, mode, cmd, assumed_family, env, abs_path), None
+        except Exception as e:
```

Comment:
This captures every exception that can be raised, while the above code silences them.